compiler/src/iree/compiler/GlobalOptimization/FuseHorizontalContractions.cpp

```diff
@@ -0,0 +1,424 @@
+    return std::nullopt;
+  }
+  Operation *user = *result.user_begin();
+  auto genericOp = dyn_cast<linalg::GenericOp>(user);
+  if (!genericOp) {
+    return std::nullopt;
+  }
+  if (genericOp.getNumDpsInputs() != 1 || genericOp.getNumDpsInits() != 1) {
+    return std::nullopt;
+  }
+  if (llvm::any_of(genericOp.getIndexingMapsArray(),
+                   [](AffineMap map) { return !map.isIdentity(); })) {
+    return std::nullopt;
+  }
+  if (genericOp.getNumParallelLoops() != genericOp.getNumLoops()) {
+    return std::nullopt;
+  }
+  auto yieldOp = cast<linalg::YieldOp>(genericOp.getBody()->getTerminator());
+  auto yieldDefOp = yieldOp->getOperand(0).getDefiningOp<arith::TruncFOp>();
+  if (!yieldDefOp) {
+    return std::nullopt;
+  }
+  auto arg = dyn_cast<BlockArgument>(yieldDefOp->getOperand(0));
+  if (!arg || arg.getParentBlock() != genericOp.getBody() ||
+      arg.getArgNumber() != 0) {
+    return std::nullopt;
+  }
+  return genericOp;
+}
+
+/// Find the operation that is dominated by all other operation.
+std::optional<Operation *>
+findDominatedByAllOps(const SetVector<Operation *> &ops,
+                      const DominanceInfo &dominanceInfo) {
+  for (auto opA : ops) {
+    bool opAIsDominatedByAll = true;
+    for (auto opB : ops) {
+      if (opA == opB) {
+        continue;
+      }
+      if (!dominanceInfo.properlyDominates(opB, opA)) {
+        opAIsDominatedByAll = false;
+        break;
+      }
+    }
+    if (opAIsDominatedByAll) {
+      return opA;
+    }
+  }
+  return std::nullopt;
+}
+
+/// Find all candidates that can be used for horizontal fusion
+static std::optional<HorizontalFusionGroup> getHorizontalFusionGroupMembers(
+    linalg::LinalgOp seedOp, std::optional<linalg::GenericOp> truncOp,
+    const llvm::SmallDenseSet<linalg::LinalgOp> &groupedOperations,
+    const DominanceInfo &dominanceInfo) {
+
+  SetVector<Operation *> allOps;
+  SmallVector<linalg::LinalgOp> contractionOps = {seedOp};
+  std::optional<SmallVector<linalg::GenericOp>> truncateOps;
+  if (truncOp) {
+    truncateOps = {truncOp.value()};
+  }
+  allOps.insert(seedOp);
+  if (truncOp) {
+    allOps.insert(truncOp.value());
+  }
+
+  Value lhs = seedOp->getOperand(0);
+  Type lhsType = lhs.getType();
+  Value rhs = seedOp->getOperand(1);
+  Type rhsType = rhs.getType();
+  Value out = seedOp->getOperand(2);
+  Type outType = out.getType();
+
+  auto canBeGrouped = [&](Operation *op) -> bool {
+    if (op->getParentOp() != seedOp->getParentOp()) {
+      return false;
+    }
+    auto linalgOp = dyn_cast<linalg::LinalgOp>(op);
+    if (!linalgOp) {
+      return false;
+    }
+    if (!linalg::isaContractionOpInterface(linalgOp)) {
+      return false;
+    }
+    if (groupedOperations.contains(linalgOp) || allOps.contains(linalgOp)) {
+      return false;
+    }
+    if (linalgOp->getOperand(0).getType() != lhsType ||
+        linalgOp->getOperand(1).getType() != rhsType ||
+        linalgOp->getOperand(2).getType() != outType) {
```

Comment:
This should also fail on dynamic shapes for the RHS right now because we can't guarantee that they are the same size for all contractions in the group (unless we want to do some analysis, but I think not worth it for now).